Documentation/technote-postgresql.md

```diff
@@ -0,0 +1,142 @@
+[![DAML logo](https://daml.com/wp-content/uploads/2020/03/logo.png)](https://www.daml.com)
+
+[![Download](https://img.shields.io/github/release/digital-asset/daml.svg?label=Download)](https://docs.daml.com/getting-started/installation.html)
+[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://github.com/digital-asset/daml/blob/master/LICENSE)
+
+# TechNote: Hardening PostgreSQL for TLS and default authentication options
+
+In this document we look at how to lock down access the PostgreSQL database connections. We use the PostgreSQL 12 Docker images as the
+basis for this configuration. It also assumes you are using the full example reference app in this repo which implements a sample
+two-tier PKI CA hierarchy.
+
+The lockdown steps includes the following:
+
+- Docker startup
+- Enable TLS certicates on the server side
+- Initialisation of the PostgreSQL database on first use
+  - initialisation scripts
+  - Create a non-superuser user and separate database
+- Enable JDBC TLS and client certs
+  - Convert keys to DER format:
+  - JDBC connection parameters
+
+## Docker startup
+
+The PostgresQL Docker image is started as follows:
+
+```aidl
+docker run --name daml-postgres -d -p 5432:5432 \
+  -e POSTGRES_PASSWORD="ChangeDefaultPassword!" \
+  -e POSTGRES_HOST_AUTH_METHOD="scram-sha-256" \
+  -e POSTGRES_INITDB_ARGS="--auth-host=scram-sha-256 --auth-local=scram-sha-256" \
+  -v "$(pwd)/certs/server/certs/db.$DOMAIN.cert.pem:/var/lib/postgresql/db.$DOMAIN.cert.pem:ro" \
+  -v "$(pwd)/certs/server/private/db.$DOMAIN.key.pem:/var/lib/postgresql/db.$DOMAIN.key.pem:ro" \
+  -v "$(pwd)/certs/intermediate/certs/ca-chain.cert.pem:/var/lib/postgresql/ca-chain.crt:ro" \
+  -v "$(pwd)/pg-initdb:/docker-entrypoint-initdb.d:ro" \
+  postgres:12 \
+  -c ssl=on \
+  -c ssl_cert_file=/var/lib/postgresql/db.$DOMAIN.cert.pem \
+  -c ssl_key_file=/var/lib/postgresql/db.$DOMAIN.key.pem \
+  -c ssl_ca_file=/var/lib/postgresql/ca-chain.crt \
+  -c ssl_min_protocol_version="TLSv1.2" \
+  -c ssl_ciphers="HIGH:!MEDIUM:+3DES:!aNULL"
+```
+
+- POSTGRES_PASSWORD # Set the default super-administrative password
+- POSTGRES_HOST_AUTH_METHOD # The default authentication method. This can be ```trust``` but this bypasses all local authentication checks
+- POSTGRES_INITDB_ARGS # To support the changed authentication method we need to change the startup parameters to the DB
+- Several -v # mount various files into the Docker file system
+- ssl=on # Enable SSL for PostgresQL
+- ssl_cert_file # The public certificate for the database TLS connection
+- ssl_key_file # private key file
+- ssl_ca_file # the CA trust chain (root and intermediate)
+- ssl_min_protocol_version="TLSv1.2" # require minimum of TLS 1.2 protocol
+- ssl_ciphers="HIGH:!MEDIUM:+3DES:!aNULL" # Refuse certain weak ciphers for TLS
+
+## PostgresQL Initialization Script
+
+The PostgresQL Docker initialization script does the following:
```

Comment:
Maybe describe where you need to save this and how it’s picked up (`-v "$(pwd)/pg-initdb:/docker-entrypoint-initdb.d:ro"`).

Reply:
Added